wcfsetup/install/files/lib/system/service/worker/Encryption.class.php

```diff
@@ -0,0 +1,194 @@
+    }
+
+    private static function addPadding(#[\SensitiveParameter] string $payload, string $contentEncoding): string
+    {
+        $length = \mb_strlen($payload, '8bit');
+        $paddingLength = ServiceWorkerHandler::MAX_PAYLOAD_LENGTH - $length;
+
+        if ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AES128GCM) {
+            return \str_pad($payload . "\x02", $paddingLength + $length, "\x00", STR_PAD_RIGHT);
+        } elseif ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AESGCM) {
+            return \pack('n*', $paddingLength) . \str_pad($payload, $paddingLength + $length, "\x00", STR_PAD_LEFT);
+        } else {
+            throw new \RuntimeException('Unknown content encoding "' . $contentEncoding . '"');
+        }
+    }
+
+    private static function createInfo(string $type, ?string $context, string $contentEncoding)
+    {
+        if ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AESGCM) {
+            \assert($context !== null);
+            \assert(\mb_strlen($context, '8bit') === 135);
+
+            return 'Content-Encoding: ' . $type . "\x00" . Encryption::CURVE_ALGORITHM . $context;
+        } elseif ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AES128GCM) {
+            return 'Content-Encoding: ' . $type . "\x00";
+        } else {
+            throw new \RuntimeException('Unknown content encoding "' . $contentEncoding . '"');
+        }
+    }
+
+    private static function getEncryptionContentCodingHeader(
+        string $contentEncoding,
+        int $length,
+        string $salt,
+        string $publicKey
+    ): string {
+        if ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AES128GCM) {
+            /** {@link https://datatracker.ietf.org/doc/html/rfc8188#section-2.1} */
+            return $salt
+                . \pack('N*', $length)
+                . \pack('C*', \mb_strlen($publicKey, '8bit'))
+                . $publicKey;
+        } elseif ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AESGCM) {
+            return "";
+        } else {
+            throw new \RuntimeException('Unknown content encoding "' . $contentEncoding . '"');
+        }
+    }
+
+    /**
+     * {@link https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-encryption-encoding-00#section-4.2}
+     */
+    private static function createContext(
+        string $clientPublicKey,
+        string $serverPublicKey,
+        string $contentEncoding
+    ): ?string {
+        if ($contentEncoding === ServiceWorker::CONTENT_ENCODING_AES128GCM) {
+            return null;
+        }
+        \assert(\mb_strlen($clientPublicKey, '8bit') === VAPID::PUBLIC_KEY_LENGTH);
+
+        $len = \pack('n', 65);
+
+        return "\x00" . $len . $clientPublicKey . $len . $serverPublicKey;
+    }
+
+    private static function getIKM(
+        ServiceWorker $serviceWorker,
+        #[\SensitiveParameter] string $sharedSecret,
+        string $userPublicKey,
+        string $newPublicKey
```

Comment:
I feel it would help to understand the code if instead of passing the serialized key, the `JWK` object would be passed, because the amount of logic in `encrypt()` would be reduced and type safety would be improved.

Reply:
However, only the serialised key is needed, not the JWK object. I do not think it is helpful to serialise the JWK object multiple times in different functions.

Reply:
Cryptographic code is hard to understand even for an experienced developer. I would recommend to write the code as misuse-resistant as possible. Leveraging the type system to assist in that is probably a good thing.

Depending on the architecture this might lead to needing to serialize the JWK multiple times, but the serialization is not really expensive, the code is not sitting in a hot path and the actual cryptographic operations and IO will likely take up the bulk of the time.